### PR TITLE
LMS Courses Update

### PR DESCRIPTION
**Pull request by dtgreiner**
## _Merging this PR_
- use the squash-merge strategy for PRs targeting a release-X branch
- use a merge-commit or rebase strategy for PRs targeting the stable branch

## Description
[//]: # (Summarize changes and include links related issue. List any new dependencies)

This updates the Talentlms integration to store the configuration (subdomains) and the courses separately. This is to prevent a bunch of unnecessary API calls as well as decouple the course and configuration to make it easier to call the API for information specific to a course and require specific courses for a user.

Screenshot 2024-10-22 at 9 55 36 AM

The concept of a default course has been added as well as the ability to override the defaults and choose specific courses to be required for a specific user. If training is required of the user, an "Edit Training" link will be available in the user admin list.

Screenshot 2024-10-22 at 10 58 16 AM

Screenshot 2024-10-22 at 10 53 28 AM

## Type of change
[//]: # (e.g., Bug fix, New feature, Code clean-up, Dependency update)
New feature

## Checklist before requesting review
[//]: # (Remove any items that are not applicable)
- [X] I have performed a self-review of my code
- [X] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [ ] If adding a new endpoint / exposing data in a new way, I have:
  - [ ] ensured the API can't leak data from other data sources
  - [ ] ensured this does not introduce N+1s
  - [ ] ensured permissions and visibility checks are performed in the right places
- [ ] I have updated the documentation (or not applicable)
- [ ] I have added spec tests (or not applicable)
- [ ] I have provided testing instructions in this PR or the related issue (or not applicable)

### CODE REVIEW

File: app/controllers/user_training_controller.rb [hunk 1]
```diff
@@ -6,39 +6,59 @@
 
 class UserTrainingController < ApplicationController
   def index
-    config = Talentlms::Config.first
-    lms = Talentlms::Facade.new
+    lms = Talentlms::Facade.new(current_user)
+    courses = current_user.required_training_courses
+    configs = [].tap do |result|
+      courses.each { |course| result << course.config }
+    end.uniq
```
Comment (eanders) [inline app/controllers/user_training_controller.rb]: Is this the same as the following?
```suggestion
    configs = courses.flat_map(&:config).uniq
```

File: app/controllers/user_training_controller.rb [hunk 2]
```diff
@@ -6,39 +6,59 @@
...
-          lms.log_course_completion(current_user, completed_on) if current_user.last_training_completed != completed_on.to_date
-          current_user.update(training_completed: true, last_training_completed: completed_on.to_date)
-          redirect_to after_sign_in_path_for(current_user)
-        else
-          # Construct TalentLMS Course URL
-          redirect_url = if current_user.can_search_own_clients?
-            clients_url
+          completed_on = lms.complete?(config, course_id)
+          if completed_on.present?
+            lms.log_course_completion(config, course_id, completed_on) # if current_user.last_training_completed != completed_on.to_date
```
Comment (eanders) [inline app/controllers/user_training_controller.rb:39]: Is this supposed to be commented out or was that from testing?

File: app/models/talentlms/course.rb [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+    # Submit a 'post' request to TalentLMS
+    #
+    # @param action [String] the REST endpoint name
+    # @param data [Hash] the post data
+    # @param args [Hash<String, String>] arguments to be added to the end of the URL
+    # @return [JSON] results
+    def post(action, data, args = nil)
+      config.post(action, data, args)
+    end
```
Comment (eanders) [inline app/models/talentlms/course.rb:42]: These are fine as is, but I _think_ you could also just say:
```
delegate :get, to: :config
delegate :post, to: :config
```

File: app/controllers/user_training_controller.rb (hunk 1 above)
Comment (dtgreiner) [reply app/controllers/user_training_controller.rb]: yes, its been added to the PR. Thanks!

**Comment by dtgreiner** [comment]
@eanders I have a new PR [here](https://github.com/greenriver/hmis-warehouse/pull/4858) created from release-139 so it doesn't have the rebase issues. I've added the recommendations already noted in this branch to the new PR

**Comment by eanders** [comment]
@dtgreiner feel free to close this if it is completely incorporated into https://github.com/greenriver/hmis-warehouse/pull/4858

**Pull request closed without merging**